tests/ui/traits/pointee-normalize-equate.rs

```diff
@@ -0,0 +1,55 @@
+// check-pass
+
+#![feature(ptr_metadata)]
+
+use std::ptr::{self, Pointee};
+
+fn cast_same_meta<T: ?Sized, U: ?Sized>(ptr: *const T) -> *const U
+where
+    T: Pointee<Metadata = <U as Pointee>::Metadata>,
+{
+    let (thin, meta) = ptr.to_raw_parts();
+    ptr::from_raw_parts(thin, meta)
+}
+
+struct Wrapper<T: ?Sized>(T);
+
+// if `Wrapper<T>: ?Sized` then normalize `Wrapper<T>::Metadata` -> `T::Metadata`
+fn wrapper_to_tail<T: ?Sized>(ptr: *const T) -> *const Wrapper<T> {
+    cast_same_meta(ptr)
+}
+
+// if `Wrapper<T>: Sized` then normalize `Wrapper<T>::Metadata` -> `()`
+fn wrapper_to_unit<T: ?Sized>(ptr: *const ()) -> *const Wrapper<T>
+where
+    Wrapper<T>: Sized,
+{
+    cast_same_meta(ptr)
+}
+
+trait Project {
+    type Assoc: ?Sized;
+}
+
+struct WrapperProject<T: ?Sized + Project>(T::Assoc);
+
+// normalize `WrapperProject<T>::Metadata` -> `T::Assoc::Metadata`
+fn wrapper_project<T: ?Sized + Project>(ptr: *const T::Assoc) -> *const WrapperProject<T> {
+    cast_same_meta(ptr)
+}
+
+// In this example, `WrapperProject<&'a T>` is `Sized` modulo regions,
+// but `?Sized` considering regions.
+// Normalize `WrapperProject<&'a T>::Metadata` -> `<&'a T as Project>::Assoc::Metadata`
+// and don't require `WrapperProject<&'a T>: Sized`.
+fn sized_modulo_regions<'a, T: ?Sized + 'static>(
```

Comment:
I think it's fine for this to result in an error. This matches the handling of where-bounds with other goals, e.g
```rust
trait Trait<'a> {}

impl<'a, T> Trait<'a> for T{}

fn impls_trait<'a, T: Trait<'a>>(_: &'a ()) {}

fn foo<'a, T: Trait<'static>>(x: &'a ()) {
    impls_trait::<T>(x)
}
```

Reply:
Alternatively, we could also always normalize from `Type::Metadata` to `Tail::Metadata` if `Type != Tail` and then make bounds like `Wrapper<T>: Sized` imply `T: Sized` (during elaboration?). That way all the cases in this test should still compile.

I will look into that,
@rustbot author

Reply:
Looks like making `Wrapper<T>: Sized` imply `T: Sized` is not straight forward, because it can break existing code by making trait selection ambiguous in a few edge cases involving higher-ranked bounds:

```rust
trait Project {
    type Assoc: ?Sized;
}

struct WrapperProject<T: ?Sized + Project>(T::Assoc);

// making (1) imply (2) causes error: "type annotations needed"
// because having both (2) and (3) in the param env makes
// `<&'a () as Project>::Assoc: Sized` ambiguous
fn foo<'a>()
where
    for<'b> &'b (): Project,
    for<'b> WrapperProject<&'b ()>: Sized, // (1)
    // for<'b> <&'b () as Project>::Assoc: Sized, // (2)
    <&'a () as Project>::Assoc: Sized, // (3)
{}
```

Because I don't know a clean solution for this problem at the moment, I've just made the `sized_modulo_regions` test case fail for now like you suggested.